source/pci/pci.go

```diff
@@ -0,0 +1,149 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package pci
+
+import (
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+	"path"
+	"strings"
+)
+
+type pciDeviceInfo map[string]string
+
+var logger = log.New(os.Stderr, "", log.LstdFlags)
+
+type NFDConfig struct {
+	DeviceClassWhitelist []string `json:"deviceClassWhitelist,omitempty"`
+	DeviceLabelFields    []string `json:"deviceLabelFields,omitempty"`
+}
+
+var Config = NFDConfig{
+	DeviceClassWhitelist: []string{"03", "0b40", "12"},
+	DeviceLabelFields:    []string{"class", "vendor"},
+}
+
+var devLabelAttrs = []string{"class", "vendor", "device", "subsystem_vendor", "subsystem_device"}
+
+// Implement FeatureSource interface
+type Source struct{}
+
+// Return name of the feature source
+func (s Source) Name() string { return "pci" }
+
+// Discover features
+func (s Source) Discover() ([]string, error) {
+	features := map[string]bool{}
+
+	devs, err := detectPci()
+	if err != nil {
+		return nil, fmt.Errorf("Failed to detect PCI devices: %s", err.Error())
+	}
+
+	// Construct a device label format, a sorted list of valid attributes
+	deviceLabelFields := []string{}
+	configLabelFields := map[string]bool{}
+	for _, field := range Config.DeviceLabelFields {
+		configLabelFields[field] = true
+	}
+
+	for _, attr := range devLabelAttrs {
+		if _, ok := configLabelFields[attr]; ok {
+			deviceLabelFields = append(deviceLabelFields, attr)
+			delete(configLabelFields, attr)
+		}
+	}
+	if len(configLabelFields) > 0 {
+		keys := []string{}
+		for key := range configLabelFields {
+			keys = append(keys, key)
+		}
+		log.Printf("WARNING: invalid fields '%v' in deviceLabelFields, ignoring...", keys)
+	}
+	if len(deviceLabelFields) == 0 {
+		log.Printf("WARNING: no valid fields in deviceLabelFields defined, using the defaults")
+		deviceLabelFields = []string{"class", "vendor"}
+	}
+
+	// Iterate over all device classes
+	for class, classDevs := range devs {
+		for _, white := range Config.DeviceClassWhitelist {
+			if strings.HasPrefix(class, strings.ToLower(white)) {
+				for _, dev := range classDevs {
+					devLabel := ""
+					for i, attr := range deviceLabelFields {
+						// Take four four first characters, so that the programming
+						// interface identifier gets stripped from the raw class code
+						devLabel += dev[attr][0:4]
```

Comment:
maybe check size before slice operation ? if for some reasons it will be shorter, it will crash with runtime error.

Reply:
All the fields are min 2 byte IDs, presented by kernel as ascii text. I think we can rely on the kernel interface not to change(?)

Reply:
Hopefully not, but if you expect to read some external data, I'd rather make sure that sizes of what was read is correct. It can be done in `readDevInfo()` then here it can be kept as is.
BTW, there in `readDevInfo()` will be also good to use `TrimPrefix()` for stripping `0x` and then do length check (4-6 symbols?)